### ISSUE TITLE
Running problem about Test_LapIRN_diff_resize.py

### ISSUE THREAD

**Issue by tzayuan** [cwmok/LapIRN#12]
Hi, @cwmok 
I'm sorry to bother you, but I have some questions which I can't sove it by myself.

I try to run `Test_LapIRN_diff_resize.py` script by using OASIS dataset, but I see that the original OASIS dataset's image size is `(160 x 192 x 224)`. Follow your instruction and code, I don't know if I should resize the above dataset for training(./Data/OASIS/) to `(144, 160, 192)` on my own before training? 
And meanwhile I found that the interpolate function resize the deformation flow to `(160 x 192 x 224)`, it bothers me because it seems represent that the size of input image should be `(160 x 192 x 224)`, but I have not found some code for resize `(160 x 192 x 224)` to `(144, 160, 192)` during the training process, so I guess that I should resize the original OASIS dataset by myself to  `(144, 160, 192)`, but this size is not same as the size of resized deformation flow, which is `(160 x 192 x 224)`.

Thanks a lot!

Yours,
Ztayuan

**Comment by cwmok** [comment]
Hi @tzayuan,

I missed two lines of code to resize the input data and I have updated the code (`Test_LapIRN_diff_resize.py`). Specifically, I added `moving_img = F.interpolate(moving_img, size=imgshape, mode='trilinear')` and `fixed_img = F.interpolate(fixed_img, size=imgshape, mode='trilinear')` to resize the input images, so you don't have to resize your dataset before using this script. Thanks for reporting the bug. :)

Best Regards,
Tony

**Comment by tzayuan** [comment]
Hi, @cwmok 

Thank you for fixing this bug!!! The logic now seems clear!!!

I have something puzzled to ask you.

I try to train this model by using `Train_LapIRN_diff.py` , my datasets' image size is `(160 x 192 x 160)`, so I modify the code anywhere the `imgshape` appear in `Train_LapIRN_diff.py`. The code can be run normally. After train finished by using `Train_LapIRN_diff.py`(lv1:30000epoch, lv2:30000epoch, lv3:60000epoch), I run the test script, but the result seems strange, but I have no idea why it happens, so can you give me some suggestions, thanks a lot!
<img width="589" alt="Snipaste_2022-04-06_13-02-53" src="https://user-images.githubusercontent.com/25738822/161899429-8e8e1f9d-61cf-4d93-93ec-45d8ad4d76c4.png">
<img width="504" alt="Snipaste_2022-04-06_13-03-15" src="https://user-images.githubusercontent.com/25738822/161899439-b43c40da-e380-44ec-8ff8-81353148cb39.png">

It looks like the image have been cropped, but I did't find the coding wrong.Thanks a lot again for your suggestions!

Yours,
Ztayuan

**Comment by cwmok** [comment]
Hi @tzayuan,

It seems that your training has collapsed. By default, `Train_LapIRN_diff.py` will not normalize your data to [0, 1] and you have to make sure that your dataset is normalized to [0, 1]. Here is my suggestion:

1. If you want to use `Train_LapIRN_diff.py`, make sure you have properly preprocessed your data (details can be found in our paper). Or, set `norm=False` for every `Dataset_epoch` in `Train_LapIRN_diff.py`.

2. Try `Train_LapIRN_diff_resize.py` if you are testing on the OASIS dataset. This example is tested and validated with the preprocessed dataset in https://github.com/adalca/medical-datasets/blob/master/neurite-oasis.md.

**Comment by tzayuan** [comment]
Hi @cwmok ,

Firstly thanks for your suggestions!!!

I tried to set `norm = True` during train and test, but after train finished, I found that the results looks still strange(the result image is included in the follow code zip):

<img width="684" alt="Snipaste_2022-04-08_10-15-35" src="https://user-images.githubusercontent.com/25738822/162349941-ad410411-c176-43a4-a1fd-cf6c4ec66b96.png">

 It seems that the image is cropped, I have no idea why it happens, so I have to seek your assistance. 

Here is my code for training a dataset which image size is `(160 * 192 * 160)`, I I made some changes but just a little, so if you could further give me some suggestions, I'm sorry to take up your time, thanks a lot!!!

I don't know why the zip file can not be loaded to Github Issue, so I transfer it into a external website:
Link：[Code Zip](https://cowtransfer.com/s/9ff412d418a54c)
Password：dqmxu9

**Comment by cwmok** [comment]
Hi @tzayuan,

I have checked your code. The issue is caused by the `Norm_Zscore` normalization in `Dataset_epoch` and the test script. Registration methods that rely on the NCC similarity function should make sure that the input images X and Y have a background intensity 0. Otherwise, the background will become the region of interest in the registration, and one of the trivial solutions is to match the background of X and Y instead of aligning the brain regions. 

Your code will be fine by removing the `Norm_Zscore` normalization in `Dataset_epoch` and the test script. You just need to make sure your input images are in [0, 1] and the background intensity of these inputs is zero (very important). Read the latest `Dataset_epoch` in https://github.com/cwmok/LapIRN/blob/master/Code/Functions.py#L111 for more details about the correct normalization method.

Feel free to let me know whether my suggestion solves your problem or not.

**Comment by tzayuan** [comment]
> Hi, @cwmok
> 
> Thank you for fixing this bug!!! The logic now seems clear!!!
> 
> I have something puzzled to ask you.
> 
> I try to train this model by using `Train_LapIRN_diff.py` , my datasets' image size is `(160 x 192 x 160)`, so I modify the code `imgshape`、`imgshape_4`、`imgshape_2` appear in `Train_LapIRN_diff.py`. The code can be run normally. After train finished by using `Train_LapIRN_diff.py`(lv1:30000epoch, lv2:30000epoch, lv3:60000epoch), I run the test script `Test_LapIRN_diff.py`, but the result seems strange, but I have no idea why it happens, so could you give me some suggestions, thanks a lot! <img alt="Snipaste_2022-04-06_13-02-53" width="589" src="https://user-images.githubusercontent.com/25738822/161899429-8e8e1f9d-61cf-4d93-93ec-45d8ad4d76c4.png"> <img alt="Snipaste_2022-04-06_13-03-15" width="504" src="https://user-images.githubusercontent.com/25738822/161899439-b43c40da-e380-44ec-8ff8-81353148cb39.png">
> 
> It looks like the image have been cropped, but I did't find the coding wrong.Thanks a lot again for your suggestions!
> 
> Yours, Ztayuan

Hi @cwmok ,

Thanks for your suggestions!!!

Before I set `norm = True`, I firstly set `norm = False` for training and test, as the result shown above. 

Here is a image which I used for training and testing the model, I have checked that the image is normalized to [0, 1], and the background's intensity value is 0. Thanks a lot!!!
image example link: [S11.delineation.skullstripped.nii.gz](https://github.com/cwmok/LapIRN/files/8449712/S11.delineation.skullstripped.nii.gz)

**Comment by cwmok** [comment]
Hi @tzayuan,

The preprocessed image looks good to me. I need more information about your issue. Could you show me your training loss of the first 2000 iterations (only level 3 is fine)? Also, could you print out the min and max value of X and Y (using torch.min(), torch.max()) before the ` F_X_Y, X_Y, Y_4x, F_xy, F_xy_lvl1, F_xy_lvl2, _ = model(X, Y)`?

It would be great if you could provide me with the `"../../../Dataset/LPBA40_delineation/delineation_l_norm/fixed.nii.gz"` as well. Then, I can test your code on my machine.

**Comment by tzayuan** [comment]
Hi @cwmok ,

Thank you very much for your suggestions!!!

I'm sorry that my loss log is missing, and I'm retraining the model without norm and keeping the loss log now.

I print out the min and max value of X and Y by using `torch.min()` `torch.max()`:
`print(torch.max(X))`
`print(torch.min(X))`
`print(torch.max(Y))`
`print(torch.min(Y))`

and the print information are as follows:
> tensor(1., device='cuda:0')
tensor(0., device='cuda:0')
tensor(1., device='cuda:0')
tensor(0., device='cuda:0')

And here is my `fixed.nii.gz`:
[fixed.nii.gz](https://github.com/cwmok/LapIRN/files/8450077/fixed.nii.gz)

Thanks a lot!!!

**Comment by cwmok** [comment]
Hi @tzayuan,

I have tested your code with your data (fixed.nii.gz and S11.delineation.skullstripped.nii.gz). Overall, the training is fine, but I spotted some minor issues in your code. I noticed that you are using an older version of our code. There are some issues/bugs as described below:

1. In the `train_lvl3()`, the code that loads the model weights for lvl2 and lvl3 is missing. Therefore, the training in `train_lvl1()` and `train_lvl2()` have no effect on the lvl3 of the model. To fix this issue, please refer to the changing log attached below.
2. The `transform_unit_flow_to_flow()` function in `Functions.py` is not updated.

Nevertheless, I cannot reproduce your result. Here are my results (trained with 2 image scans, 1000 iterations, lvl3 only, norm=False).
Fixed image (fixed.nii.gz):
![fixed](https://user-images.githubusercontent.com/63316551/162484184-4b2e0c93-8a08-4be7-8d9a-fec0a20639f8.png)
Moving image (S11.delineation.skullstripped.nii.gz)
![moving](https://user-images.githubusercontent.com/63316551/162484227-4ea73002-27e0-41ca-bca5-20fe9807fb56.png)
Warped Moving image (warpped_moving.nii.gz)
on.skullstripped.nii.gz):
![warped](https://user-images.githubusercontent.com/63316551/162484284-a5aaac4e-121f-4a78-8d71-9f0d78fd837c.png)

For your reference, here is the configuration of my machine:
OS: Windows 10
GPU: Nvidia Titan RTX
Pytorch: torch 1.10.0+cu113
Python: Pythjon 3.8 (Anaconda3)

I have also attached the changing log, running log and source code for your reference.
Code: [test_Code.zip](https://github.com/cwmok/LapIRN/files/8453876/test_Code.zip)
Logs:  [Changing_and_running_log.txt](https://github.com/cwmok/LapIRN/files/8453893/Changing_and_running_log.txt)

If you have any idea/clue about the problem, feel free to let me know. I am more than happy to help.

**Comment by tzayuan** [comment]
Hi, @cwmok 

Sorry for disturbing you again!

I referred to the code you provided and completed 3 levels train, but the result seems still strange, I have checked possible reason, but I can't solve it on myself, so I have to ask for your closer help again. 

Here's my part visualized results:
1) when I finished level1(30000epoch)、level2(30000epoch)、level3(1000epoch), the final warped image is as follows:
![image](https://user-images.githubusercontent.com/25738822/163175851-fa55fcba-db14-47dd-8223-d3c76360cdc8.png)
Left is warped image, right is fixed image

2) when I finished level1(30000epoch)、level2(30000epoch)、level3(3000epoch), the final warped image is as follows:
![image](https://user-images.githubusercontent.com/25738822/163176142-17476518-19ce-4575-8e82-01f410e84f60.png)
Left is warped image, right is fixed image

3) when I finished level1(30000epoch)、level2(30000epoch)、level3(5000epoch), the final warped image is as follows:
![image](https://user-images.githubusercontent.com/25738822/163176317-5c119b7f-3683-4e22-a34c-fd75dfb69515.png)
Left is warped image, right is fixed image

My total training loss log:[logdebugall.log](https://github.com/cwmok/LapIRN/files/8481262/logdebugall.log)

Code for train & test:[LapIRN_0413.zip](https://github.com/cwmok/LapIRN/files/8481277/LapIRN_0413.zip)

In addition, I follow your suggestion that set the environ at bash level, I think it is a pretty good habit for sharing code and debug the code. Thanks a lot!!!

Best regards, 
Tzayuan

**Comment by cwmok** [comment]
Hi @tzayuan,

It is weird. Have you tried the "(Example) Training on the preprocessed OASIS dataset with downsampled images" we provided? Are this example works in your machine? 

Recently, I received some similar issues in my other GitHub repository ([issue1](https://github.com/cwmok/Fast-Symmetric-Diffeomorphic-Image-Registration-with-Convolutional-Neural-Networks/issues/10), [issue2](https://github.com/cwmok/LapIRN/issues/11)). It seems that people running our code on their server with multiple GPU may face unexpected issues. Since I don't have a dedicated GPU server, I cannot reproduce the results and figure out the reason.

One solution reported in [issue1](https://github.com/cwmok/Fast-Symmetric-Diffeomorphic-Image-Registration-with-Convolutional-Neural-Networks/issues/10) is to repack our code using Docker. Yet, the reason for your issue is still undergoing investigation. 

I suggest you try the "(Example) Training on the preprocessed OASIS dataset with downsampled images" first. To make sure that we are using the exact same codes and training dataset. If the problem persists, I think it is related to some non-trivial environmental issue and I need your help to figure it out.

Regards,
Tony

**Comment by tzayuan** [comment]
Hi, @cwmok 

Thank you for your enthusiastic help!

I'm sorry for the slight delay due to the lack of computing resources, I will take your advice and test the example code first and give you feedback as soon as I finish.

**Comment by cwmok** [comment]
Hi @tzayuan,

I look forward to hearing from you.

For the consultation of PhD application, you may send me an email directly (cwmokab {at} connect {dot} ust {dot} hk). I am glad to share my experience with you.

Regards,
Tony

**Comment by tzayuan** [comment]
Hi @cwmok ,

I'm training the LapIRN model by using resized OASIS dataset now.

And In the waiting process, I read the source code, I found that in the training code part, it seems that the whole dataset is sent to the network for training, i.e. there's no prepared test data? As `names = sorted(glob.glob(datapath + '/OASIS_OAS1_*_MR1/aligned_norm.nii.gz'))`.

If should I split the dataset on myself before training?

Thanks a lot!

Best regard, 
Ztayuan

**Comment by cwmok** [comment]
>> If should I split the dataset on myself before training?

Yes, you should split the dataset into train, test, validation sets by yourself.

>> In addition, I load the training model of lvl3 epoch40000 and run Test_LapIRN_diff_resize.py, but the model seems have something wrong, and the error print is as follow, I'm also examining the cause of the error now.
Error log:[error.txt](https://github.com/cwmok/LapIRN/files/8541768/error.txt)

If you are using the default parameters for training, you will train a model with "--start_channel" = 6. Therefore, when using the `Test_LapIRN_diff_resize.py`, set "--start_channel" = 6 and set "--modelpath" to your modal will solve your problem .
